core/src/files.rs

```diff
@@ -0,0 +1,212 @@
+use std::{
+    ffi::{OsStr, OsString},
+    path::PathBuf,
+    rc::Rc,
+};
+
+use codespan::ByteIndex;
+use codespan_reporting::files::Error;
+use nickel_lang_vector::Vector;
+
+use crate::{position::RawSpan, stdlib::StdlibModule};
+
+#[derive(
+    Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, serde::Serialize, serde::Deserialize,
+)]
+pub struct FileId(u32);
+
+#[derive(Debug, Clone)]
+pub struct File {
+    /// The name of the file.
+    name: OsString,
+    /// The source code of the file.
+    source: Rc<str>,
+    /// The starting byte indices in the source code.
+    line_starts: Rc<[ByteIndex]>,
```

Comment:
I'm not sure what does `line_starts` represent, even with the comment.

Reply:
Hopefully the changed one is clearer:
```
    /// The byte index of the start of each line. The first element of this array is always 0.
```